worker/lease/store.go

```diff
@@ -0,0 +1,422 @@
+
+	if len(keys) == 1 {
+		q += `
+WHERE  t.type = ?
+AND    l.model_uuid = ?
+AND    l.name = ?`
+
+		name = "LeasesForKey"
+		key := keys[0]
+		args = []any{key.Namespace, key.ModelUUID, key.Lease}
+	}
+
+	stmt, err := s.getPrepared(context.Background(), name, q)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rows, err := stmt.Query(args...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	result, err := leasesFromRows(rows)
+	return result, errors.Trace(rows.Err())
+}
+
+// ClaimLease (lease.Store) claims the lease indicated by the input key,
+// for the holder and duration indicated by the input request.
+// The lease must not already be held, otherwise an error is returned.
+func (s *Store) ClaimLease(key lease.Key, req lease.Request, stop <-chan struct{}) error {
+	if err := req.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+INSERT INTO lease (uuid, lease_type_id, model_uuid, name, holder, start, expiry)
+SELECT ?, id, ?, ?, ?, datetime('now'), datetime('now', ?) 
+FROM   lease_type
+WHERE  type = ?`[1:]
+
+		stmt, err := s.getPrepared(ctx, "ClaimLease", q)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		d := fmt.Sprintf("+%d seconds", int64(math.Ceil(req.Duration.Seconds())))
+
+		_, err = stmt.ExecContext(
+			ctx, utils.MustNewUUID().String(), key.ModelUUID, key.Lease, req.Holder, d, key.Namespace)
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.Trace(<-errCh)
+	case err := <-errCh:
+		cancel()
+		if database.IsErrConstraintUnique(err) {
+			return lease.ErrHeld
+		}
+		return errors.Trace(err)
+	}
+}
+
+// ExtendLease (lease.Store) ensures the input lease will be held for at least
+// the requested duration starting from now.
+// If the input holder does not currently hold the lease, an error is returned.
+func (s *Store) ExtendLease(key lease.Key, req lease.Request, stop <-chan struct{}) error {
+	if err := req.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+UPDATE lease
+SET    expiry = datetime('now', ?)
+WHERE  uuid = (
+    SELECT l.uuid
+    FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+    WHERE  t.type = ?
+    AND    l.model_uuid = ?
+    AND    l.name = ?
+    AND    l.holder = ?
+)`[1:]
+
+		stmt, err := s.getPrepared(ctx, "ExtendLease", q)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		d := fmt.Sprintf("+%d seconds", int64(math.Ceil(req.Duration.Seconds())))
+
+		result, err := stmt.ExecContext(ctx, d, key.Namespace, key.ModelUUID, key.Lease, req.Holder)
+
+		// If no rows were affected, then either this key does not exist or
+		// it is not held by the input holder, constituting an invalid request.
+		if err == nil {
+			var affected int64
+			affected, err = result.RowsAffected()
```

Comment:
FYI: 

> If a separate thread makes changes on the same database connection while sqlite3_changes() is running then the value returned is unpredictable and not meaningful.

https://www.sqlite.org/c3ref/changes.html

Reply:
This might not necessarily be the mechanism for populating `RowsAffected`. LXD appear to use it in [a few places](https://github.com/lxc/lxd/search?p=1&q=RowsAffected).